src/ripple/app/paths/PathRequest.cpp

```diff
@@ -442,92 +441,141 @@ void PathRequest::resetLevel (int l)
 {
-    STPath fullLiquidityPath;
-    auto result = fp.findPathsForIssue(issue,
-        mContext[issue], fullLiquidityPath);
-    if (! result)
+    auto i = currency_map.find(currency);
+    if (i != currency_map.end())
+        return i->second;
+    auto pathfinder = std::make_unique<Pathfinder>(
+        cache, *raSrcAccount, *raDstAccount, currency,
+            boost::none, dst_amount, saSendMax);
+    if (pathfinder->findPaths(level))
+        pathfinder->computePathRanks(max_paths_);
+    else
+        pathfinder.reset();  // It's a bad request - clear it.
+    return currency_map[currency] = std::move(pathfinder);
+}
+
+void
+PathRequest::findPaths (RippleLineCache::ref cache, int const level,
+    Json::Value& jvArray)
+{
+    auto sourceCurrencies = sciSourceCurrencies;
+    if (sourceCurrencies.empty ())
     {
-        m_journal.debug << iIdentifier << " No paths found";
-        return false;
+        auto usCurrencies =
+            accountSourceCurrencies(*raSrcAccount, cache, true);
+        bool sameAccount = *raSrcAccount == *raDstAccount;
+        for (auto const& c : usCurrencies)
+        {
+            if (!sameAccount || (c != saDstAmount.getCurrency()))
+            {
+                if (c.isZero())
+                    sourceCurrencies.insert({ c, xrpAccount() });
+                else
+                    sourceCurrencies.insert({ c, *raSrcAccount });
+            }
+        }
     }
-    mContext[issue] = *result;
 
-    boost::optional<PaymentSandbox> sandbox;
-    sandbox.emplace(&*cache->getLedger(), tapNONE);
+    auto const dst_amount = convert_all_ ?
+        STAmount(saDstAmount.issue(), STAmount::cMaxValue, STAmount::cMaxOffset)
+            : saDstAmount;
+    hash_map<Currency, std::unique_ptr<Pathfinder>> currency_map;
+    for (auto const& issue : sourceCurrencies)
+    {
+        JLOG(m_journal.debug)
+            << iIdentifier
+            << " Trying to find paths: "
+            << STAmount(issue, 1).getFullText();
+
+        auto& pathfinder = getPathFinder(cache, currency_map,
+            issue.currency, dst_amount, level);
+        if (! pathfinder)
+        {
+            assert(false);
+            m_journal.debug << iIdentifier << " No paths found";
+            continue;
+        }
 
-    auto& sourceAccount = ! isXRP(issue.account)
-        ? issue.account
-        : isXRP(issue.currency)
-        ? xrpAccount()
-        : *raSrcAccount;
+        STPath fullLiquidityPath;
+        auto ps = pathfinder->getBestPaths(max_paths_,
+            fullLiquidityPath, mContext[issue], issue.account);
+        mContext[issue] = ps;
 
-    STAmount saMaxAmount = saSendMax.value_or(
-        STAmount({issue.currency, sourceAccount}, 1u, 0, true));
+        boost::optional<PaymentSandbox> sandbox;
+        sandbox.emplace(&*cache->getLedger(), tapNONE);
 
-    m_journal.debug << iIdentifier
-        << " Paths found, calling rippleCalc";
+        auto& sourceAccount = ! isXRP(issue.account)
```

Comment:
I first saw this code a year ago - and yet everytime I see it again after a few weeks, I have to decipher what it's doing again.  And the worst is that I'm not sure I could explain it to others, I merely convince myself that it's "obvious".

Perhaps putting comments inline is a bad idea - perhaps instead there needs to be a half-page somewhere explaining exactly what's going on - but I feel this would be much better with some sort of higher-level explanation of the process.